diesel_codegen/src/insertable.rs

```diff
@@ -15,19 +17,176 @@ pub fn derive_insertable(item: syn::MacroInput) -> quote::Tokens {
+                  #value_type: diesel::insertable::InsertValues<DB>,
+        {
+            type Values = #value_type;
+
+            #value_fn
+        }
+    )
+}
+
+fn make_column_insert_value_definition(
+    a: &Attr,
+    table_name: &syn::Ident,
+    struct_name: &str
+) -> quote::Tokens {
+    let column_name = a.column_name
+        .as_ref()
+        .expect(&format!("Unknown column name while implementing Insertable for {}"
+                         , struct_name));
+    let field_ty = &a.ty;
+    quote!(
+        diesel::insertable::ColumnInsertValue<
+            #table_name::#column_name,
+            diesel::expression::helper_types::AsNullableExpr<
+                &'insert #field_ty,
+                #table_name::#column_name,
+            >
+        >)
+}
+
+fn make_column_insert_value(
+    a: &Attr,
+    table_name: &syn::Ident,
+    struct_name: &str
+) -> quote::Tokens {
+    let column_name = a.column_name
+        .as_ref()
+        .expect(&format!("Unknown column name while implementing Insertable for {}"
+                         , struct_name));
+    let column = quote!(#table_name::#column_name);
+    let field_access = a.name_for_pattern();
+
+    match a.field_kind() {
+        "option" => {
+            quote!(
+                match #field_access {
+                    value @ &Some(_) => {
+                        ColumnInsertValue::Expression(
+                            #column,
+                            AsExpression::<<<#column as Expression>::SqlType
+                                as IntoNullable>::Nullable>::as_expression(value)
+                        )
+                    },
+                    &None => ColumnInsertValue::Default(#column),
+                }
+            )
+        }
+        "regular" => {
+            quote!(
+                ColumnInsertValue::Expression(
+                    #column,
+                    AsExpression::<<<#column as Expression>::SqlType
+                        as IntoNullable>::Nullable>::as_expression(#field_access)
+                )
+            )
+        },
+        _ => panic!("Unknown field kind while implementing Insertable for {}", struct_name),
+    }
+}
+
+fn generate_impl_into_insert_statement(model: &Model) -> quote::Tokens {
+    let insert = syn::LifetimeDef::new("'insert");
+    let generics = syn::aster::from_generics(model.generics.clone())
+        .with_lifetime(insert.clone())
+        .ty_param_id("Op")
+        .ty_param_id("Ret")
+        .with_predicates(model.generics.lifetimes
+                         .iter()
+                         .map(|l|{
+                             syn::WherePredicate::RegionPredicate(
+                                 syn::WhereRegionPredicate{
+                                     lifetime: l.lifetime.clone(),
+                                     bounds: vec![insert.lifetime.clone()]
+                                 }
+                             )
+                         }))
+        .build();
```

Comment:
Can you use block ident? I'd write this as

```rust
…
.with_predicates(
    model.generics.lifetimes
    .iter()
    .map(|l|{
        syn::WherePredicate::RegionPredicate(
            syn::WhereRegionPredicate{
                lifetime: l.lifetime.clone(),
                bounds: vec![insert.lifetime.clone()]
            }
        )
    })
)
.build();
```

(Also note space before `{` and trailing comma in `WhereRegionPredicate`)